File: main/community/src/main/kotlin/nodove/com/community/domain/Post.kt
package nodove.com.community.domain

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime

@Entity
@Table(name = "posts", indexes = [
    Index(name = "idx_author_id", columnList = "authorId"),
    Index(name = "idx_category_id", columnList = "category_id"),
    Index(name = "idx_created_at", columnList = "createdAt")
])
data class Post(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(nullable = false, length = 500)
    var title: String,

    @Column(nullable = false, columnDefinition = "TEXT")
    var content: String,

    @Column(length = 1000)
    var excerpt: String? = null,

    @Column(nullable = false)
    val authorId: String, // Auth Service의 User ID 참조

    @Column(nullable = false)
    val authorNickname: String, // 캐싱용

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    var category: Category,

    @ElementCollection
    @CollectionTable(name = "post_tags", joinColumns = [JoinColumn(name = "post_id")])
    @Column(name = "tag", length = 50)
    var tags: MutableList<String> = mutableListOf(),

    @Column(nullable = false)
    var viewCount: Int = 0,

    @Column(nullable = false)
    var likeCount: Int = 0,

    @CreationTimestamp
    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime? = null,

    @UpdateTimestamp
    @Column(nullable = false)
    val updatedAt: LocalDateTime? = null,

    @Column
    var deletedAt: LocalDateTime? = null
) {
    fun incrementViewCount() {
        viewCount++
    }

    fun incrementLikeCount() {
        likeCount++
    }

    fun decrementLikeCount() {
        if (likeCount > 0) {
            likeCount--
        }
    }

    fun softDelete() {
        deletedAt = LocalDateTime.now()
    }
}

File: flutter_app/lib/features/posts/domain/models/post.dart
import 'package:json_annotation/json_annotation.dart';
import '../../../auth/domain/models/user.dart';

part 'post.g.dart';

@JsonSerializable()
class Post {
  final int id;
  final String title;
  final String content;
  final String? excerpt;
  final User author;
  final String category;
  final List<String> tags;
  final int viewCount;
  final int likeCount;
  final int commentCount;
  final bool isLiked;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deletedAt;

  const Post({
    required this.id,
    required this.title,
    required this.content,
    this.excerpt,
    required this.author,
    required this.category,
    required this.tags,
    required this.viewCount,
    required this.likeCount,
    this.commentCount = 0,
    required this.isLiked,
    required this.createdAt,
    required this.updatedAt,
    this.deletedAt,
  });

  factory Post.fromJson(Map<String, dynamic> json) => _$PostFromJson(json);
  Map<String, dynamic> toJson() => _$PostToJson(this);

  Post copyWith({
    int? id,
    String? title,
    String? content,
    String? excerpt,
    User? author,
    String? category,
    List<String>? tags,
    int? viewCount,
    int? likeCount,
    int? commentCount,
    bool? isLiked,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? deletedAt,
  }) {
    return Post(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      excerpt: excerpt ?? this.excerpt,
      author: author ?? this.author,
      category: category ?? this.category,
      tags: tags ?? this.tags,
      viewCount: viewCount ?? this.viewCount,
      likeCount: likeCount ?? this.likeCount,
      commentCount: commentCount ?? this.commentCount,
      isLiked: isLiked ?? this.isLiked,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      deletedAt: deletedAt ?? this.deletedAt,
    );
  }
}
